AMM/project/scripts/swap-exact-input/src/main.sw

```diff
@@ -0,0 +1,58 @@
+script;
+
+use libraries::{AMM, Exchange};
+use std::block::height;
+
+enum InputError {
+    RouteTooShort: (),
+    SwapAmountsNotExact: (),
+}
+
+enum SwapError {
+    PairExchangeNotRegistered: (ContractId, ContractId),
+}
+
+fn main(
+    amm_contract_address: ContractId,
+    assets: Vec<ContractId>,
+    amounts: Vec<u64>,
```

Comment:
Is there a way to make this function just take a single input amount? The output of one swap should always be exactly the input of the next, otherwise the user will end up with small amounts of intermediate tokens (if the swap price is better than expected), or failing swaps (if the price swap price is worse than expected).

You could compare the balance of the sender before and after each swap, and then use the delta as the input amount for the next swap. 

e.g. For swap A -> B -> C -> D etc.

User provides input amount of token A, `a`. Script gets the pre-swap balance of token B,  `b`. Swap A->B occurs. Script gets new balance `b'`, and sets input amount for the B->C swap to `(b' - b)`. Script gets pre-swap balance of token C, `c`. Swap B->C occurs. Script gets new balance `c'`, and sets input amount for the C->D swap to `(c' - c)`. .. .etc.